### PR TITLE
Add WebID+Delegate support for TLS and OIDC modes

### PR DESCRIPTION
**Pull request by smalinin**
Added support of WebID+Delegate for both auth modes TLS and OIDC
You must have already WebID certificate installed in you browser for use `Register`,
(if you don't have WebID client certificate, you could create it free on http://ods-qa.openlinksw.com/youid/ or etc)
when new account is created, the server write Delegator profile to `card` file.
The card file URL (ex <https://myaccount.solid.openlinksw.com:8443/profile/card#me> ) must be added as relation to your WebID profile like:
```
@prefix ns1:	<http://id.myopenlink.net/DAV/home/smalinin/YouID/Sergey_Malinin_Fb_id_myopenlink_net/profile.ttl#> .

ns1:identity_cert:identity <http://www.openlinksw.com/schemas/cert#onBehalfOf>  <https://myaccount.solid.openlinksw.com:8443/profile/card#me>  .
```
The created Delegator profile may be used in browser with YouID extension https://github.com/OpenLinkSoftware/youid  for Login to Solid-Server.

### CODE REVIEW

**Comment by dmitrizagidulin** [comment]
@smalinin - Thank you for this PR! Couple of style questions, before we dive into the meat of the code.
Does this PR pass the unit tests and the StandardJS linter?
I noticed that there were several sections that don't match Standard conventions. Such as the use of snake case variable and function names (for example, `handle_register_webid`) -- we only use camel case. Similarly, I noticed semicolons in several places -- Standard is strictly no-semicolon.

**Comment by smalinin** [comment]
I will prepared the update for this pull request. Now I have an additional fixes, will fix the style and etc.

**Comment by melvincarvalho** [comment]
Running 

    standard --fix

Should help fix a few errors

**Comment by RubenVerborgh** [comment]
> Does this PR pass the unit tests and the StandardJS linter?

But more importantly, why didn't Travis catch this?
Did someone switch off Travis integration, or disable it for pull requests?

Worked until 5 days ago: https://travis-ci.org/solid/node-solid-server/pull_requests

**Comment by dmitrizagidulin** [comment]
@RubenVerborgh - I was wondering that same thing, why didn't CI catch it! Does it maybe have something to do with the PR coming in from a forked repo?

**Comment by RubenVerborgh** [comment]
PR is a PR, should work. Maybe it will magically jump back into action after the conflict is resolved.

**Comment by dmitrizagidulin** [comment]
Ah, good point, maybe that’s it

**Comment by smalinin** [comment]
I updated code (all last updates have been pushed):
 - style updated for camelCase
 - remove semicolons
 - all tests are passed now ( I updated 3 tests )
 - the code used rdfstore.js was replaced with new code used rdflib.js
all last updates in https://github.com/OpenLinkSoftware/node-solid-server/commit/fd0efa71d22fe0decd8064fdf36f51b6a68b6985   

Must I create new pull request or ?

**Comment by melvincarvalho** [comment]
@smalinin that's great, no need for new PR

but the package-lock.json has conflicts

I think this must be resolved before the automated tests kick in (ie travis)

**Comment by smalinin** [comment]
The conflict with package-lock.json may be because I was need to install global-tunnel-ng package.
It could not work with default packages on my host, after the simple `npm install`
```
npm install -s global-tunnel-ng

```

**Comment by smalinin** [comment]
@melvincarvalho

**Comment by melvincarvalho** [comment]
you possibly need to check in favicon.ico

**Comment by smalinin** [comment]
Fixed package-lock.json

**Comment by melvincarvalho** [comment]
The tests passed, excellent!

And thanks also for adding new tests.

Looks reasonable at first blush.  Although it would be good to know a summary of what has changed.  Aside from the middleware accepting delegation, there are changes to the registration and login forms I think.

I suggest we merge this and put it live so that it can be tested.  

Any reviews from @RubenVerborgh @dmitrizagidulin @timbl @ericprud welcome!

**Comment by melvincarvalho** [comment]
@smalinin please feel free to drop into chat : 

https://gitter.im/solid/node-solid-server

If you'd prefer to discuss changes in realtime

**Comment by smalinin** [comment]
@melvincarvalho
If I run:

File: default-templates/new-account/index.html [hunk 1]
```diff
@@ -3,7 +3,7 @@
 <head>
   <meta charset="utf-8">
   <meta name="viewport" content="width=device-width, initial-scale=1">
-  <title>{{#if name}}{{name}} &ndash; {{/if}}Solid Home</title>
+  <title>Personal Data Space (or POD) {{#if name}} of {{name}} {{/if}}</title>
```
Comment (RubenVerborgh) [inline default-templates/new-account/index.html]: These changes are unrelated to this PR.

File: default-templates/new-account/index.html [hunk 2]
```diff
@@ -16,9 +16,13 @@
...
     <div class="container-fluid">
       <div class="navbar-header">
-        <a href="/" class="navbar-brand">{{#if name}}{{name}} &ndash; {{/if}}Solid Home</a>
+        <a href="/" class="navbar-brand">Personal Data Space (POD)</a>
       </div>
       <div class="navbar-right">
+        <a id="logged-href" href="" title="" style="display:none">
+          <span class="glyphicon glyphicon-user"></span>
+        </a>
+        <span>&nbsp&nbsp</span>
```
Comment (RubenVerborgh) [inline default-templates/new-account/index.html]: Incorrect HTML.

File: default-templates/new-account/index.html [hunk 3]
```diff
@@ -16,9 +16,13 @@
   <nav class="navbar navbar-default">
     <div class="container-fluid">
       <div class="navbar-header">
-        <a href="/" class="navbar-brand">{{#if name}}{{name}} &ndash; {{/if}}Solid Home</a>
+        <a href="/" class="navbar-brand">Personal Data Space (POD)</a>
```
Comment (RubenVerborgh) [inline default-templates/new-account/index.html]: These changes are unrelated to this PR.

File: default-templates/new-account/index.html [hunk 4]
```diff
@@ -16,9 +16,13 @@
...
       </div>
       <div class="navbar-right">
+        <a id="logged-href" href="" title="" style="display:none">
+          <span class="glyphicon glyphicon-user"></span>
+        </a>
```
Comment (RubenVerborgh) [inline default-templates/new-account/index.html]: Why is this needed?

File: default-templates/new-account/index.html [hunk 5]
```diff
@@ -29,12 +33,28 @@
     <div class="col-md-12">
       <p class="lead">
         <span>
-          This is a public homepage of {{#if name}}{{name}}, whose WebID is{{else}}a user with WebID{{/if}}
+          This is a public homepage of <a href="{{webId}}">{{#if name}} {{name}} {{/if}}</a>
```
Comment (RubenVerborgh) [inline default-templates/new-account/index.html]: Why is the WebID gone?

File: default-templates/new-account/index.html [hunk 6]
```diff
@@ -29,12 +33,28 @@
...
+        </span>
+      </p>
+    </div>
+  </div>
+
+{{#if tlsCertWebid}}
+  <div class="row" id="update_profile" style="display:none">
+    <div class="col-md-12">
+      <p class="lead">
+        Regarding WebID-TLS+Delegation protocol, add the following RDF statement to your Delegates WebID-Profile Document or click "Update WebID-Profile Doc" to attempt this action right now.
```
Comment (RubenVerborgh) [inline default-templates/new-account/index.html]: Is this fixed in the template?

File: default-templates/new-account/profile/card [hunk 7]
```diff
@@ -23,3 +28,25 @@
...
+    owl:sameAs <{{extWebId}}> .
+{{/if}}
+
+{{#if tlsCertWebid}}
+:me
+    acl:delegates <{{tlsCertWebid}}> ; # Delegation Relations for Software Agent 
+    cert:key [
+               a cert:RSAPublicKey ;
+               cert:exponent   "65537"^^xsd:int ;
+               cert:modulus    "00"^^xsd:hexBinary
```
Comment (melvincarvalho) [inline default-templates/new-account/profile/card]: Just checking did you want to put "00" in here or {{tlsCertModulus}} ?

File: default-templates/new-account/index.html (hunk 2 above)
Comment (melvincarvalho) [inline default-templates/new-account/index.html]: possible semi colon here

File: default-templates/new-account/profile/card (hunk 7 above)
Comment (smalinin) [reply default-templates/new-account/profile/card]: Yes, now `card` profile is create as Delegator profile(that doesn't need PublicKey), so it is something like a stub only for properly work(loading data) with browser YouID extension.

File: default-templates/new-account/index.html [hunk 8]
```diff
@@ -122,17 +151,100 @@ <h1>Apps</h1>
...
+      logged_href.href = "";
+      logged_href.title = "";
     } else {
       actionButton.innerText = 'Log out'
+      if (session.webId) {
+        logged_href.style.display = 'initial'
+        logged_href.href = session.webId;
+        logged_href.title = session.webId;
+      }
     }
```
Comment (RubenVerborgh) [inline default-templates/new-account/index.html]: A bit too much magic here. Why not separate Log In and Log Out buttons that we show and hide, instead of one multipurpose button?

File: default-templates/new-account/index.html [hunk 9]
```diff
@@ -122,17 +151,100 @@ <h1>Apps</h1>
...
   }
 
   function sessionActionButton() {
     return document.getElementById('session-action')
   }
+  function sessionLogged_href() {
```
Comment (RubenVerborgh) [inline default-templates/new-account/index.html]: Naming is off.

File: default-templates/new-account/index.html [hunk 10]
```diff
@@ -122,17 +151,100 @@ <h1>Apps</h1>
...
+              'Content-Type': 'text/xml; charset=utf-8'
+            },
+            credentials: 'include', 
+            body: msg
+        })
+        .then(function(response) {
+          alert('Attr: '+response.statusText);
+        })
+        .catch(function(error) {})
+***/
```
Comment (RubenVerborgh) [inline default-templates/new-account/index.html]: Please remove.

File: default-templates/new-account/profile/card (hunk 7 above)
Comment (RubenVerborgh) [reply default-templates/new-account/profile/card]: That doesn't seem proper usage of modulus then though. Is there any way around this?

File: default-templates/new-account/index.html (hunk 2 above)
Comment (smalinin) [reply default-templates/new-account/index.html]: Fixed

File: default-templates/new-account/profile/card (hunk 7 above)
Comment (smalinin) [reply default-templates/new-account/profile/card]: Yes, it is possible avoid this, I will fix our browser YouID extension

File: default-templates/new-account/index.html (hunk 2 above)
Comment (smalinin) [reply default-templates/new-account/index.html]: Fixed

File: default-templates/new-account/index.html (hunk 9 above)
Comment (smalinin) [reply default-templates/new-account/index.html]: Commented code was removed

**Comment by smalinin** [comment]
I will prepare the Readme for created changes.

**Comment by RubenVerborgh** [comment]
What are all the favicon changes about?

**Comment by smalinin** [comment]
@RubenVerborgh 
The favicon.ico(.acl) wasn't in the server root https://localhost  before, it was only in the user home https://user.localhost

**Comment by RubenVerborgh** [comment]
I see. We might need to consider splitting this PR in multiple ones, as it seems to address multiple concerns.

File: default-templates/new-account/index.html (hunk 1 above)
Comment (kidehen) [reply default-templates/new-account/index.html]: So can't you just remove the index.html then?

File: lib/models/account-template.js [hunk 11]
```diff
@@ -79,10 +79,20 @@ class AccountTemplate {
    * @return {Object}
    */
   static templateSubstitutionsFor (userAccount) {
+    let addOwlSameAs = userAccount.externalWebId &&
+           userAccount.tls_certificate_webid &&
+           userAccount.externalWebId !== userAccount.tls_certificate_webid
```
Comment (melvincarvalho) [inline lib/models/account-template.js]: Is this part correct.  Should it be camel case?

Comment (smalinin) [reply lib/models/account-template.js]: Yes, it is a bug, I missed this rename, the fix has been commited

File: default-views/auth/login-tls.hbs [hunk 12]
```diff
@@ -1,36 +1,92 @@
...
+            authClientStore = {}
+          }
+          authClientStore.session = session
+          localStorage.setItem(authClientNamespace, JSON.stringify(authClientStore))
+
+          response.text().then(function(text) {
+            try {
+              const url = new URL(text);
+              window.location.href = url.href;
+            } catch(err) {
```
Comment (RubenVerborgh) [inline default-views/auth/login-tls.hbs]: What can go wrong here?

File: lib/models/authenticator.js [hunk 13]
```diff
@@ -192,11 +196,40 @@ class TlsAuthenticator extends Authenticator {
...
+    if (this.authMethod === 'tls' && !this.username) {
+      throw new Error('Username required')
+    }
+
+    if (this.username) {
+      return this.accountManager.accountExists(this.username)
+        .then(exists => {
+          if (!exists) {
+            debug(`Account ${this.username} is not existed`)
+            throw new Error(`Account ${this.username} is not existed`)
```
Comment (RubenVerborgh) [inline lib/models/authenticator.js]: * does not exist

File: lib/models/authenticator.js [hunk 14]
```diff
@@ -293,10 +329,54 @@ class TlsAuthenticator extends Authenticator {
...
-  verifyWebId (certificate, callback) {
+  verifyWebId (certificate, authCallback) {
     debug('Verifying WebID URI')
 
-    webid.verify(certificate, callback)
+    var me = this
+    var delegator
+
+    if (me.headers) {
+      delegator = me.headers['on-behalf-of'] || me.headers['On-Behalf-Of'] || null
```
Comment (RubenVerborgh) [inline lib/models/authenticator.js]: Should only need to be one case (lowercase).

File: lib/webid_util.js [hunk 15]
```diff
@@ -0,0 +1,266 @@
...
+          } else {
+            rc = kb.any(webIdUri, kb.sym('http://www.w3.org/2004/02/skos/core#altLabel'))
+            if (rc) {
+              name = rc.value
+            }
+          }
+        }
+      }
+    }
+    return {name}
```
Comment (RubenVerborgh) [inline lib/webid_util.js]: We should avoid this pyramid code, that's hard to maintain. We could for instance put all those predicates in an array and do a reduce.

File: lib/create-server.js [hunk 16]
```diff
@@ -28,6 +29,10 @@ function createServer (argv, app) {
     delete argv.idp
   }
 
+  if (argv.httpProxy) {
+    globalTunnel.initialize(argv.httpProxy)
```
Comment (RubenVerborgh) [inline lib/create-server.js]: We should document this config option.

File: default-templates/new-account/index.html (hunk 1 above)
Comment (smalinin) [reply default-templates/new-account/index.html]: Rollback this fix

File: default-templates/new-account/index.html (hunk 3 above)
Comment (smalinin) [reply default-templates/new-account/index.html]: Rollback this fix

File: default-templates/new-account/index.html (hunk 4 above)
Comment (smalinin) [reply default-templates/new-account/index.html]: It is for show user icon as href=webId, now you don't see your webId after the Login.

File: default-templates/new-account/index.html (hunk 6 above)
Comment (smalinin) [reply default-templates/new-account/index.html]: Rollback fix to original version

File: default-templates/new-account/index.html (hunk 5 above)
Comment (smalinin) [reply default-templates/new-account/index.html]: Rollback fix to original version

File: default-templates/new-account/index.html (hunk 6 above)
Comment (smalinin) [reply default-templates/new-account/index.html]: What do you mean ? The file default-templates/new-account/index.html is template for user's index.html

File: default-templates/new-account/index.html (hunk 8 above)
Comment (smalinin) [reply default-templates/new-account/index.html]: The "magic" was added for show user icon(like anchor with href=webId)  on the left side of Login/Logout button.  The code for Login/Logout button was already in the original version.

**Comment by smalinin** [comment]
About updates in PR (I hope, I didn't forgot anything)

- updated WebID-TLS and WebID-OIDC auth for support WebID+Delegate-TLS and WebID+Delegate-OIDC

- index.html was updated:
    -  show user icon(like anchor with href=webId) on the left side of Login/Logout button.
    - added button "Update WebID-Profile doc" for add next Delegator relation
```
<TLS-WebID-URI>   <http://www.openlinksw.com/schemas/cert#onBehalfOf>   <https://user.localhost:8444/profile/card#me> .
```

- Auth OIDC
    - Register:
        - added loading TLS WebID or DelegatorWebID as default value to field "External WebID"
        - the "card" file is updated for using it as Delegator profile
        - if ExternalWebID was a Delegator, the next relation is added to all .acl files
       `<#owner> acl:agent  <ExternalWebID> .`
    - Login:
        - Username filed could not be empty now for both UID/PWD Login and WebID+Delegate-TLS Login. It is needed for propely redirect to userHome,  if ExternalWebID is used for Login
       
- Auth TLS
    - Register was not implemented before
       - now it create "card" profile, that could be used as Delegator profile
       - the UI field "Name" is loaded by default with value from TLS WebId(Certificate.CN)  or Delegator (name from profile data)
       - the UI field "WebID" is loaded by default with TLS WebID or Delegator WebID
       - if ExternalWebID was a Delegator, the next relation is added to all .acl files
       `<#owner> acl:agent  <ExternalWebID> .`
   - Login was not implemented before
        - Username filed could not be empty now for both UID/PWD Login and WebID+Delegate-TLS Login. It is needed for propely redirect to userHome, if ExternalWebID is used for Login

File: lib/webid_util.js (hunk 15 above)
Comment (smalinin) [reply lib/webid_util.js]: Fixed

**Comment by melvincarvalho** [comment]
@smalinin changelog looks great, thanks! 

you might want to add a line about fixing the favicon too

**Comment by RubenVerborgh** [comment]
@smalinin Do we have a README though with a scenario?

**Comment by melvincarvalho** [comment]
@smalinin good progress!  

Do you have this branch (4.1) running on a live server?

It would be good to try and get it running so we can test live, and in parallel explain how it can be tested.

**Comment by kidehen** [comment]
@melvincarvalho, 

Our live servers are:
1. https://solid.openlinksw.com:8443 -- WebID-TLS+Delegation IdP
1. https://solid.openlinksw.com:8444 -- WebID-OIDC + WebID-TLS+Delegation Idp

Right now, there are live instances from this latest release at:

1. https://kidehen3.solid.openlinksw.com:8443
1. https://kidehen5.solid.openlinksw.com:8444

**Comment by smalinin** [comment]
I am sorry, but I was need to add back to removed fix, that was mark as outdated,
because in authMode = OIDC, the Login with Certificate(WebID+Delegate TLS) doesn't work .
The Server return back redirect command and WebID in HTTP Header **User**, when server authMode== OIDC, the HTTP Header **User** is removed by browser from HTTP Response Headers. I don't have idea, why it is happened, because the same things works properly when server authMode == TLS.  
Therefore I added fix back to sources. 
Now server doesn't sent back HTTP 302 redirect, it returns back the redirect URL as String, the client code receives this string and execute redirect.

**Comment by dmitrizagidulin** [comment]
> in authMode = OIDC, the Login with Certificate(WebID+Delegate TLS) doesn't work

Could part of the issue be - when authMode = OIDC, the Express app doesn't request the TLS cert from the connection? Or are you passing in the cert in the cert header somehow?

File: default-templates/new-account/public/.acl [hunk 17]
```diff
@@ -6,6 +6,9 @@
 <#owner>
     a acl:Authorization;
     acl:agent <{{webId}}>;
+{{#if addOwlSameAs}}
```
[the remark this answers is not in the document]
Comment (smalinin) [reply default-templates/new-account/public/.acl]: If you Register new account with externalWebID from Delegator, 
the code from **{{#if addOwlSameAs}}** is added. 
If you Register new account with externalWebID from TLS Certificate, it isn't added.
p.s.
It is done, because I could not add owl:sameAs relation for externalWebID(that is the Delegator also). **owl:sameAs** isn't supported by rdflib.js

**Comment by smalinin** [comment]
@dmitrizagidulin 
The browser ask me the TLS certificate, when I click on button **[With Certificate (WebID+Delegate-TLS)]** in Login page.

File: default-templates/new-account/public/.acl (hunk 17 above)
[the remark this answers is not in the document]
Comment (smalinin) [reply default-templates/new-account/public/.acl]: I added the next relation to **card** profile (if externalWebID is from Delegator)
```
:me  owl:sameAs <${externalWebID}> .
```

**Comment by melvincarvalho** [comment]
@smalinin Thanks for this awesome contribution and for responding quickly to feedback!

Spoke to @RubenVerborgh today and we agreed to merge this into the v4.1.0 branch and test it further.

I think it's possible to add more changes.

Our hope is to then merge 4.1 into master and upgrade the version.

**Comment by melvincarvalho** [comment]
Is it possible to leave this issue open, as it's not yet been merged into master?

Or shall we perhaps start a fresh issue?

**Comment by melvincarvalho** [comment]
Ah It's a PR.  I think we should associate an issue with it?

File: default-templates/new-account/settings/.acl [hunk 18]
```diff
@@ -6,6 +6,9 @@
 
     acl:agent
         <{{webId}}>;
+{{#if addOwlSameAs}}
```
Comment (dmitrizagidulin) [inline default-templates/new-account/settings/.acl]: I still think this section does not need to be added to *all* the .acls

File: lib/webid_util.js [hunk 19]
```diff
@@ -0,0 +1,255 @@
...
+
+exports.createWebIdUtils = function () {
+  return new WebIdUtils()
+}
+
+/**
+ * A class wrapping utility routines for WebID handling.
+ * @constructor
+ */
+function WebIdUtils () {
```
Comment (dmitrizagidulin) [inline lib/webid_util.js]: Would it make sense to refactor this to use ES6 `class` syntax, to match the rest of the code base?
